### PR TITLE
refactor(app): Permit door open status during certain recovery steps

### PR DESCRIPTION
**Pull request by mjhuff**
Closes [EXEC-700](https://opentrons.atlassian.net/browse/EXEC-700)

# Overview

Currently, when a user is on an error recovery step that contains copy like "do XYZ on the deck", opening the robot door causes the "door is open" modal to appear, which is a bit annoying when trying to read directions.

While there are quite a few ways to solve this issue, in practice this requires a bit of thought when considering app takeovers, ODD/desktop app syncing, and operating within error recovery's current routing structure. This PR solves the above through a new `RECOVERY_MAP_METADATA` object, in which all steps specifically permit the door to be open or not. If allowed, we don't show the modal on that step, otherwise we do. Typescript enforces that any new route and steps added must specify whether the door is permitted to be open on that step. Yes, this may seem annoying, but I do think we want developers to think through whether the door should be open on every step.

To handle robot commands that a user may initiate on a "door is permitted open" view, in which cases we absolutely DO want the door to be closed, we add to the wrapper that occurs before commands (now called `handleMotionRouting`) that ensures the door is closed, routing to the `RecoveryDoorOpen` modal if necessary and rejecting the command. After closing the door, the user is routed back to the view right before command execution.

In effect, there are now two circumstances which cause `RecoveryDoorOpen` to appear:

1. The old but adjusted behavior - the door is opened when it shouldn't be, and we render the door modal _instead of_ the recovery content. When the door status is ok, we just render the recovery content. There is no change in the actual route.
2. Entirely new behavior - we manually adjust the route to the door modal and then route back to where we after the door status is resolved. In other words, the door open modal is recovery content. We do this when handling robot commands, as explained above.

While this may sound complicated, in practice it has little bearing. If we ever do multi-app syncing (same place in ER at the same time), this plays nicely with it.

Lastly, this PR adds a cleanup util that resets state when another app takes over the recovery session. This is more pertinent now that door routing needs knowledge of the previous route, and we should reset this and some other state whenever a takeover happens.

TLDR: As always, door logic is special-cased and requires special considerations.

## Test Plan and Hands on Testing
- Smoke tested error recovery with the new door changes.

## Changelog
- Opening the door during error recovery on steps with directions no longer shows the door open warning.

## Risk assessment
low-mediumish. 

[EXEC-700]: https://opentrons.atlassian.net/browse/EXEC-700?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by sfoster1 — APPROVED** [review]
I think this is a really nice way to handle this, thanks!
